src/pykeen/utils.py

```diff
@@ -1447,6 +1447,98 @@ def ensure_complex(*xs: torch.Tensor) -> Iterable[torch.Tensor]:
+    each nodes' neighborhood, and hashes these multi-sets into a new set of colors.
+
+    .. note::
+        more precisely, this implementation calculates the results of the 1-Weisfeiler-Lehman algorithm. There is a
+        hierarchy of k-WL tests, which color k-tuples of nodes instead of single nodes
+
+    .. note::
+        Graph Neural Networks are closely related to the 1-WL test, cf. e.g., https://arxiv.org/abs/1810.02244
+
+    .. note::
+        colorings based on the WL test have been used to define graph kernels, cf., e.g.,
+        https://www.jmlr.org/papers/volume12/shervashidze11a/shervashidze11a.pdf
+
+    .. note::
+        the implementation creates intermediate dense tensors of shape `(num_nodes, num_colors)`
+
+    .. seealso::
+        https://towardsdatascience.com/expressive-power-of-graph-neural-networks-and-the-weisefeiler-lehman-test-b883db3c7c49
+
+    :param edge_index: shape: `(2, m)`
+        the edge list
+    :param max_iter:
+        the maximum number of iterations
+    :param num_nodes:
+        the number of nodes. If None, will be inferred from the edge index.
+
+    :raises ValueError:
+        if the number of nodes exceeds `torch.long` (this cannot happen in practice, as the edge index tensor
+        construction would already fail earlier)
+
+    :yields: the colors for each Weisfeiler-Lehman iteration
+    """
+    num_nodes = num_nodes or edge_index.max().item() + 1
+    colors = edge_index.new_zeros(size=(num_nodes,), dtype=torch.long)
+    # note: in theory, we could return this uniform coloring as the first coloring; however, for featurization,
+    #       this is rather useless
+
+    # initial: degree
+    # note: we calculate this separately, since we can use a more efficient implementation for the first step
+    unique, counts = edge_index.unique(return_counts=True)
+    colors[unique] = counts
+
+    # hash
+    colors = colors.unique(return_inverse=True)[1]
+    yield colors
+
+    # determine small integer type for dense count array
+    for idtype in (torch.int8, torch.int16, torch.int32, torch.int64):
+        if torch.iinfo(idtype).max >= num_nodes:
+            dense_dtype = idtype
+            logger.debug(f"Selected dense dtype: {dense_dtype}")
+            break
+    else:
+        raise ValueError(f"{num_nodes} too large")
+
+    adj = torch.sparse_coo_tensor(
+        indices=edge_index, values=torch.ones(size=edge_index[0].shape), device=edge_index.device
+    )
+    for _ in range(max_iter - 1):
+        # message passing: collect colors of neighbors
+        # dense colors: shape: (n, c)
+        # adj:          shape: (n, n)
+        color_sparse = torch.sparse_coo_tensor(
+            indices=torch.stack([torch.arange(num_nodes, device=colors.device), colors], dim=0),
+            # values need to be float, since torch.sparse.mm does not support integer dtypes
+            values=torch.ones_like(colors, dtype=torch.get_default_dtype()),
```

Comment:
device?

Reply:
`torch.ones_like` takes care of it. I have to provide the dtype, since it is different than the one of colors.